File: examples_py/examples_py/controllers/mpcc_controller.py
#!/usr/bin/env python3

# Copyright 2025 Universidad Politécnica de Madrid
# SPDX-License-Identifier: BSD-3-Clause

"""Acados MPCC (Model Predictive Contouring Control) adapter.

Self-contained: the controller loads the mission YAML inside
:meth:`initialize` and builds its own arc-length-reparametrised spline.
``compute_command()`` ignores the ``references`` argument and instead
sets the solver's online parameters from a sliding window over the
internal spline. This mirrors the way
``mpc_acados_position.MpcPositionController`` cooks its own per-stage
references from a single goal pose — the controller "knows" the path,
the framework just provides the per-tick tick.

Required reference fields are reported as ``POSITION`` (the minimum
that lets ``WaypointsSimulator`` validate the controller/generator
match); whatever the paired generator emits is discarded by this
adapter.
"""

__authors__ = 'Rafael Perez-Segui'
__license__ = 'BSD-3-Clause'

import math
import os
import sys
import tempfile
import time
from dataclasses import dataclass
from pathlib import Path
from typing import List, Optional

import numpy as np
import yaml
from mavpy.model import State

from examples_py.framework import (
    ControlCommand,
    ExampleConfig,
    IController,
    ReferenceField,
    ReferenceSample,
    make_mask,
)
from examples_py.generators.circuit_generator import CircuitGenerator
from examples_py.utils import mission_loader as ml


def _import_mpcc_acados():
    """Locate the mpcc_acados Python bindings."""
    try:
        from mpcc_acados import MPC
        from mpcc_acados.utils.mpc_yaml import configure_mpc_from_yaml
        return MPC, configure_mpc_from_yaml
    except ImportError:
        pass
    candidates = []
    env_dir = os.environ.get('MPCC_ACADOS_EXAMPLE_DIR')
    if env_dir:
        candidates.append(Path(env_dir).parent)  # mpcc_acados is one level above
    repo_root = os.environ.get('MPCC_REPO_ROOT')
    if repo_root:
        candidates.append(Path(repo_root) / 'workspace/thirdparty_libs/mpcc')
    candidates.append(Path('/home/rafa/mpcc_v2/workspace/thirdparty_libs/mpcc'))
    for cand in candidates:
        if (cand / 'mpcc_acados' / '__init__.py').is_file():
            sys.path.insert(0, str(cand))
            from mpcc_acados import MPC
            from mpcc_acados.utils.mpc_yaml import configure_mpc_from_yaml
            return MPC, configure_mpc_from_yaml
    raise ImportError(
        'MpccController: cannot locate mpcc_acados. Set MPCC_REPO_ROOT or '
        'MPCC_ACADOS_EXAMPLE_DIR so the loader can find the mpcc_acados '
        'Python package.')


_MPCC_SPLINE_KNOTS = 20
_SOFT_WEIGHT_DEFAULT = 2.0
_SOFT_WEIGHT_ORIGINAL = 1.0
_SEARCH_RADIUS = 1.0      # m around the previous s_eval
_DS_COARSE = 0.10
_DS_FINE = 0.01


@dataclass
class MpccControllerConfig:
    """Parsed configuration for :class:`MpccController`.

    Three groups of fields:
      * ``mpc_yaml_path``: file with the MPCC weights (``gains_*``,
        ``softcontraints_*``, ``lbu``, ``ubu``) consumed by
        :func:`mpcc_acados.utils.mpc_yaml.configure_mpc_from_yaml`.
      * ``ocp_json_file_path``: the acados-generated OCP json (defaults
        to the one shipped next to ``mpc_yaml_path``).
      * Mission paths + cruise tuning, shared with
        :class:`CircuitGenerator` (the controller cooks its own spline,
        independent of any generator).
    """

    mpc_yaml_path: str = ''
    ocp_json_file_path: str = ''
    mission_yaml: str = ''
    gates_yaml: str = ''
    desired_speed: float = 4.0
    origin_offset_m: float = 3.0
    closing_exit_margin_m: float = 3.0
    target_segment_length: float = 1.0
    samples_per_segment: int = 400


def _find_closest_s(drone_pos: np.ndarray, window, evaluate_fn,
                    s_min: float, s_max: float) -> float:
    """Coarse + fine search for the closest spline point to drone_pos."""
    best_s = s_min
    min_d = float('inf')
    s = s_min
    while s <= s_max:
        pt = evaluate_fn(s, window)
        d = math.dist(drone_pos, pt.position[:3])
        if d < min_d:
            min_d = d
            best_s = s
        s += _DS_COARSE
    s_fine = max(s_min, best_s - _DS_COARSE)
    s_end = min(s_max, best_s + _DS_COARSE)
    while s_fine <= s_end:
        pt = evaluate_fn(s_fine, window)
        d = math.dist(drone_pos, pt.position[:3])
        if d < min_d:
            min_d = d
            best_s = s_fine
        s_fine += _DS_FINE
    return best_s


class MpccController(IController):
    """Self-contained MPCC adapter (ignores the framework's references)."""

    def __init__(self, cfg: MpccControllerConfig) -> None:
        if not cfg.mpc_yaml_path:
            raise ValueError('MpccController: mpc_yaml_path must be provided.')
        if cfg.desired_speed <= 0.0:
            raise ValueError('MpccController: desired_speed must be > 0.')
        self._cfg = cfg
        self._mpc = None
        self._mpc_data = None
        self._traj = None
        self._evaluate_fn = None
        self._setpoint_cls = None
        self._traj_cls = None
        self._s_eval = 0.0
        self._control_period = 0.04
        self._last_solve_us = 0.0
        self._last_desired_velocity = np.zeros(3, dtype=float)
        self._name = 'MpccController'

    # ------------------------------------------------------------------
    # YAML config
    # ------------------------------------------------------------------

    @staticmethod
    def load_config_from_yaml(path: str) -> MpccControllerConfig:
        """Read ``configs/controllers/config_mpcc.yaml``.

        Top-level keys:
          * ``controller.ocp_json_file_path`` (optional) — acados OCP path.
          * ``mpcc.parameters`` / ``mpcc.constraints`` — solver weights and
            bounds. Uses the ``mpcc:`` key (NOT ``mpc:``) so the file is
            visually distinct from `config_mpc_trajectory.yaml`. The
            adapter rewrites the block to ``mpc:`` at runtime when
            handing it to the upstream ``configure_mpc_from_yaml``
            loader (which is hard-coded against ``mpc:``).
          * ``circuit`` (optional) — mission anchors mirroring
            :class:`CircuitGenerator`'s config.
        """
        if not os.path.isfile(path):
            raise ValueError(f'Config file not found at {os.path.abspath(path)}.')
        with open(path, 'r') as f:
            root = yaml.safe_load(f) or {}
        if not isinstance(root, dict):
            raise ValueError('mpcc config: root must be a mapping.')
        if 'mpcc' not in root:
            raise ValueError(
                f"mpcc config {path}: missing top-level 'mpcc:' block. "
                "(If you are porting an old config, rename `mpc:` -> `mpcc:`.)")
        cfg = MpccControllerConfig(mpc_yaml_path=path)
        ctrl = root.get('controller')
        if isinstance(ctrl, dict):
            if 'ocp_json_file_path' in ctrl:
                cfg.ocp_json_file_path = str(ctrl['ocp_json_file_path'])
        circuit = root.get('circuit')
        if isinstance(circuit, dict):
            for k in ('mission_yaml', 'gates_yaml'):
                if k in circuit:
                    setattr(cfg, k, str(circuit[k]))
            for k in ('desired_speed', 'origin_offset_m', 'closing_exit_margin_m',
                      'target_segment_length'):
                if k in circuit:
                    setattr(cfg, k, float(circuit[k]))
            if 'samples_per_segment' in circuit:
                cfg.samples_per_segment = int(circuit['samples_per_segment'])
        return cfg

    # ------------------------------------------------------------------
    # IController
    # ------------------------------------------------------------------

    def initialize(self, initial_state: State, example_cfg: ExampleConfig) -> None:
        if example_cfg.mpc_dt <= 0.0:
            raise ValueError('MpccController: example_cfg.mpc_dt must be > 0.')

        MPC, configure_mpc_from_yaml = _import_mpcc_acados()
        # MPCC solver --------------------------------------------------
        from examples_py.generators.circuit_generator import _import_spline_py
        self._traj_cls, self._setpoint_cls, self._evaluate_fn = _import_spline_py()

        ocp_json = self._cfg.ocp_json_file_path
        if not ocp_json:
            # Fall back, in order of preference:
            #   1. ``MPCC_ACADOS_EXAMPLE_DIR/mpcc_interface/...`` (env override
            #      set by the bench harness when the package is co-located
            #      with the mpcc repo).
            #   2. ``${MPCC_REPO_ROOT}/workspace/thirdparty_libs/mpcc/...``
            #      (canonical mpcc_v2 layout).
            #   3. Sibling of the mpc_yaml_path (the upstream mpcc layout).
            for cand_dir in [
                os.environ.get('MPCC_ACADOS_EXAMPLE_DIR'),
                (os.path.join(os.environ['MPCC_REPO_ROOT'],
                              'workspace/thirdparty_libs/mpcc/'
                              'mpcc_acados_example')
                 if os.environ.get('MPCC_REPO_ROOT') else None),
                str(Path(self._cfg.mpc_yaml_path).resolve().parent.parent),
            ]:
                if not cand_dir:
                    continue
                cand = os.path.join(cand_dir, 'mpcc_interface',
                                    'mpc_generated_code', 'acados_ocp.json')
                if os.path.isfile(cand):
                    ocp_json = cand
                    break
        if not ocp_json or not os.path.isfile(ocp_json):
            raise ValueError(
                f'MpccController: acados ocp_json_file_path not found '
                f'(tried explicit + MPCC_ACADOS_EXAMPLE_DIR + '
                f'MPCC_REPO_ROOT + sibling of {self._cfg.mpc_yaml_path}). '
                f'Set MPCC_ACADOS_EXAMPLE_DIR or specify the path in '
                f'configs/controllers/config_mpcc.yaml under '
                f'controller.ocp_json_file_path.')
        self._mpc = MPC(ocp_json_file=ocp_json)
        # `configure_mpc_from_yaml` upstream is hardcoded to read the
        # `mpc:` top-level key, but our config files use `mpcc:` so
        # they don't collide with `config_mpc_trajectory.yaml`. Rewrite
        # the block to a tmp YAML with the legacy key before invoking
        # the loader, then drop the temp file.
        configure_mpc_from_yaml(self._mpc, self._rekey_mpcc_to_mpc(
            self._cfg.mpc_yaml_path))
        self._mpc_data = self._mpc.get_data()
        self._control_period = float(self._mpc.get_prediction_time_step())
        if example_cfg.mpc_dt > 0:
            # Honour the framework's mpc_dt as the outer-loop step. The
            # MPCC solver still uses its own prediction step internally.
            self._control_period = float(example_cfg.mpc_dt)

        # Mission + spline --------------------------------------------
        # Path precedence: env var > config YAML > package demo defaults.
        # The env vars let downstream consumers (e.g. mpcc_v2's
        # launchers/simulation/launch/mav.sh) inject their own mission
        # without editing the generic configs shipped here.
        env_mission = os.environ.get('CIRCUIT_MISSION_YAML', '')
        env_gates = os.environ.get('CIRCUIT_GATES_YAML', '')
        ml_mission, ml_gates = ml._default_paths()  # noqa: SLF001
        mission_yaml = env_mission or self._cfg.mission_yaml or ml_mission
        gates_yaml = env_gates or self._cfg.gates_yaml or ml_gates

        # Match the CircuitGenerator's anchor logic: post-takeoff pose
        # (initial.xy, takeoff_altitude_m) so the MPCC's first solve
        # sees zero contour error throughout the takeoff phase. See
        # examples_py/generators/circuit_generator.py for rationale.
        origin_pose = np.asarray(initial_state.position, dtype=float).copy()
        takeoff_alt = float(getattr(example_cfg, 'takeoff_altitude_m', 0.0))
        if takeoff_alt > 0.0:
            origin_pose[2] = takeoff_alt

        setpoints = CircuitGenerator.build_setpoints(
            initial_position=origin_pose,
            mission_yaml=mission_yaml,
            gates_yaml=gates_yaml,
            origin_offset_m=self._cfg.origin_offset_m,
            closing_exit_margin_m=self._cfg.closing_exit_margin_m,
            setpoint_cls=self._setpoint_cls,
        )
        self._traj = self._traj_cls(
            setpoints,
            _MPCC_SPLINE_KNOTS,
            self._cfg.target_segment_length,
            self._cfg.samples_per_segment,
        )
        self._s_eval = 0.0
        self._last_solve_us = 0.0
        self._last_desired_velocity = np.zeros(3, dtype=float)

    def reference_horizon_size(self) -> int:
        # MPCC builds the horizon internally from the spline. The framework
        # still queries 1 sample per tick so the generator hook fires.
        return 1

    def reference_horizon_dt(self) -> float:
        return self._control_period

    def control_period(self) -> float:
        return self._control_period

    def required_reference_fields(self) -> ReferenceField:
        # We do not consume the reference samples — POSITION is the
        # cheapest mask that satisfies WaypointsSimulator's compatibility
        # check against any of the position / trajectory generators.
        return make_mask([ReferenceField.POSITION])

    def name(self) -> str:
        return self._name

    def last_solve_time_micros(self) -> float:
        return self._last_solve_us

    def last_desired_velocity(self) -> np.ndarray:
        return self._last_desired_velocity

    def provides_desired_velocity(self) -> bool:
        return True

    # ------------------------------------------------------------------
    # YAML re-key shim
    # ------------------------------------------------------------------

    @staticmethod
    def _rekey_mpcc_to_mpc(source_path: str) -> str:
        """Write a temporary copy of ``source_path`` with the top-level
        ``mpcc:`` block renamed to ``mpc:`` so the upstream
        :func:`configure_mpc_from_yaml` (which is hardcoded against
        ``mpc:``) can consume it. The temp file lives in ``$TMPDIR``
        and is left for the OS to reap — `configure_mpc_from_yaml` only
        reads it once at construction time."""
        with open(source_path, 'r', encoding='utf-8') as f:
            root = yaml.safe_load(f) or {}
        if not isinstance(root, dict) or 'mpcc' not in root:
            return source_path
        rekeyed = dict(root)
        rekeyed['mpc'] = rekeyed.pop('mpcc')
        fd, tmp_path = tempfile.mkstemp(prefix='mpcc_rekey_', suffix='.yaml')
        with os.fdopen(fd, 'w', encoding='utf-8') as f:
            f.write('# AUTO-GENERATED by MpccController._rekey_mpcc_to_mpc — '
                    'do not edit by hand.\n')
            yaml.safe_dump(rekeyed, f, sort_keys=False, default_flow_style=None)
        return tmp_path

    # ------------------------------------------------------------------
    # MPCC step
    # ------------------------------------------------------------------

    def compute_command(self, state: State,
                         references: List[ReferenceSample]) -> ControlCommand:
        del references  # MPCC sources its references from the internal spline
        if self._mpc is None or self._traj is None:
            raise RuntimeError('MpccController: initialize() must be called first.')

        pos = np.asarray(state.position, dtype=float)
        q = np.asarray(state.orientation, dtype=float)
        vel = np.asarray(state.linear_velocity, dtype=float)

        # MPCC state vector: [x, y, z, qw, qx, qy, qz, vx, vy, vz, theta]
        mpc_state = np.array([pos[0], pos[1], pos[2],
                              q[0], q[1], q[2], q[3],
                              vel[0], vel[1], vel[2],
                              self._s_eval], dtype=float)
        self._mpc_data.state.vector = mpc_state

        window, s_eval = self._traj.get_trajectory_window(self._s_eval)
        self._s_eval = s_eval

        params = self._mpc_data.parameters
        params.set_spline_points(
            np.asarray(window.path.position[:3 * _MPCC_SPLINE_KNOTS], dtype=float))
        params.set_spline_tangents(
            np.asarray(window.path.tangent[:3 * _MPCC_SPLINE_KNOTS], dtype=float))
        params.set_spline_face_points(
            np.asarray(window.facing_points[:3 * _MPCC_SPLINE_KNOTS], dtype=float))
        params.set_s_lengths(
            np.asarray(window.s_values[:_MPCC_SPLINE_KNOTS], dtype=float))
        soft = np.full(_MPCC_SPLINE_KNOTS, _SOFT_WEIGHT_DEFAULT, dtype=float)
        for og in window.original_points:
            if 0 <= og.index < _MPCC_SPLINE_KNOTS:
                soft[og.index] = _SOFT_WEIGHT_ORIGINAL
        params.set_spline_softconstraints(soft)

        t0 = time.perf_counter()
        try:
            self._mpc.solve()
        except Exception as exc:
            raise RuntimeError(f'MpccController: solver failed: {exc}') from exc
        self._last_solve_us = (time.perf_counter() - t0) * 1e6

        actu = np.asarray(self._mpc_data.actuation.vector, dtype=float)
        thrust = float(actu[0])
        rates = actu[1:4].astype(float).copy()

        # Update s_eval by projecting the drone onto the spline (monotonic).
        s_top = float(window.s_values[-1])
        s_min = max(0.0, self._s_eval - _SEARCH_RADIUS)
        s_max = min(s_top, self._s_eval + _SEARCH_RADIUS)
        best_s = _find_closest_s(pos, window, self._evaluate_fn, s_min, s_max)
        best_s = max(self._s_eval, best_s)
        self._s_eval = min(max(best_s, 0.0), s_top)

        # Expose stage-1 predicted velocity (best proxy for v_des).
        self._last_desired_velocity = self._cfg.desired_speed * np.asarray(
            window.path.tangent[:3], dtype=float)

        return ControlCommand(thrust_n=thrust, angular_rate=rates)

File: examples_cpp/src/controllers/mpcc_controller.cpp
// Copyright 2025 Universidad Politécnica de Madrid
// SPDX-License-Identifier: BSD-3-Clause

/**
 * @file mpcc_controller.cpp
 * @author Rafael Perez-Segui <r.psegui@upm.es>
 */

#include "controllers/mpcc_controller.hpp"

#include <yaml-cpp/yaml.h>
#include <unistd.h>

#include <algorithm>
#include <array>
#include <chrono>
#include <cmath>
#include <cstdlib>
#include <cstdio>
#include <filesystem>
#include <fstream>
#include <limits>
#include <stdexcept>
#include <vector>

#include "generators/circuit_generator.hpp"
#include "mpcc_acados/acados_mpc_datatype.hpp"
#include "mpcc_acados/acados_mpc_yaml.hpp"
#include "utils/example_config_utils.hpp"
#include "utils/mission_loader.hpp"

namespace mpc_examples::adapters {

namespace ml = mpc_examples::utils::mission_loader;

namespace {

constexpr int    kMpccSplineKnots   = 20;
constexpr double kSoftWeightDefault = 2.0;
constexpr double kSoftWeightOg      = 1.0;
constexpr double kSearchRadiusM     = 1.0;
constexpr double kDsCoarse          = 0.10;
constexpr double kDsFine            = 0.01;

std::string envOrEmpty(const char* name) {
  const char* v = std::getenv(name);
  return (v != nullptr) ? std::string(v) : std::string();
}

/// Auto-locate the acados OCP json shipped alongside the MPCC YAML.
std::string resolveOcpJson(const std::string& explicit_path,
                           const std::string& mpc_yaml_path) {
  if (!explicit_path.empty() && std::filesystem::exists(explicit_path)) {
    return explicit_path;
  }
  std::vector<std::string> candidate_dirs;
  if (const char* d = std::getenv("MPCC_ACADOS_EXAMPLE_DIR")) {
    candidate_dirs.emplace_back(d);
  }
  if (const char* root = std::getenv("MPCC_REPO_ROOT")) {
    candidate_dirs.emplace_back(std::string(root) +
                                "/workspace/thirdparty_libs/mpcc/mpcc_acados_example");
  }
  // Sibling of the mpc_yaml_path (upstream mpcc layout, two levels up).
  candidate_dirs.push_back(
      std::filesystem::absolute(mpc_yaml_path).parent_path().parent_path().string());
  for (const auto& cand_dir : candidate_dirs) {
    if (cand_dir.empty()) continue;
    const std::filesystem::path cand =
        std::filesystem::path(cand_dir) / "mpcc_interface" / "mpc_generated_code" /
        "acados_ocp.json";
    if (std::filesystem::exists(cand)) {
      return cand.string();
    }
  }
  throw std::invalid_argument(
      "MpccController: cannot locate acados ocp_json_file_path (tried explicit + "
      "MPCC_ACADOS_EXAMPLE_DIR + MPCC_REPO_ROOT + sibling of " + mpc_yaml_path + ").");
}

/// Coarse + fine search for the spline arc-length closest to @p drone_pos.
double findClosestS(const Eigen::Vector3d& drone_pos,
                    const spline::arc_length_reparam::ArcLengthReparametrization& window,
                    double s_min, double s_max) {
  double best_s = s_min;
  double min_d  = std::numeric_limits<double>::infinity();
  for (double s = s_min; s <= s_max; s += kDsCoarse) {
    const auto ev = spline::arc_length_reparam::evaluateArcLengthSpline(s, window);
    if (ev.position.size() < 3) continue;
    const double dx = ev.position[0] - drone_pos.x();
    const double dy = ev.position[1] - drone_pos.y();
    const double dz = ev.position[2] - drone_pos.z();
    const double d  = std::sqrt(dx * dx + dy * dy + dz * dz);
    if (d < min_d) {
      min_d  = d;
      best_s = s;
    }
  }
  const double s_lo = std::max(s_min, best_s - kDsCoarse);
  const double s_hi = std::min(s_max, best_s + kDsCoarse);
  for (double s = s_lo; s <= s_hi; s += kDsFine) {
    const auto ev = spline::arc_length_reparam::evaluateArcLengthSpline(s, window);
    if (ev.position.size() < 3) continue;
    const double dx = ev.position[0] - drone_pos.x();
    const double dy = ev.position[1] - drone_pos.y();
    const double dz = ev.position[2] - drone_pos.z();
    const double d  = std::sqrt(dx * dx + dy * dy + dz * dz);
    if (d < min_d) {
      min_d  = d;
      best_s = s;
    }
  }
  return best_s;
}

}  // namespace

MpccController::MpccController(const Config& cfg) : cfg_(cfg) {
  if (cfg_.mpc_yaml_path.empty()) {
    throw std::invalid_argument("MpccController: mpc_yaml_path must be provided.");
  }
  if (cfg_.desired_speed <= 0.0) {
    throw std::invalid_argument("MpccController: desired_speed must be > 0.");
  }
}

MpccController::Config MpccController::loadConfigFromYaml(const std::string& path) {
  const YAML::Node root = mpc_examples::detail::loadYamlRoot(path);
  if (!root["mpcc"]) {
    throw std::invalid_argument(
        "mpcc config " + path +
        ": missing top-level 'mpcc:' block. (If porting an old config, rename `mpc:` -> `mpcc:`.)");
  }
  Config cfg;
  cfg.mpc_yaml_path = path;

  if (const YAML::Node ctrl = root["controller"]; ctrl && ctrl.IsMap()) {
    if (ctrl["ocp_json_file_path"]) {
      cfg.ocp_json_file_path = ctrl["ocp_json_file_path"].as<std::string>();
    }
  }
  if (const YAML::Node circuit = root["circuit"]; circuit && circuit.IsMap()) {
    if (circuit["mission_yaml"]) cfg.mission_yaml = circuit["mission_yaml"].as<std::string>();
    if (circuit["gates_yaml"])   cfg.gates_yaml   = circuit["gates_yaml"].as<std::string>();
    if (circuit["desired_speed"]) {
      cfg.desired_speed = circuit["desired_speed"].as<double>();
    }
    if (circuit["origin_offset_m"]) {
      cfg.origin_offset_m = circuit["origin_offset_m"].as<double>();
    }
    if (circuit["closing_exit_margin_m"]) {
      cfg.closing_exit_margin_m = circuit["closing_exit_margin_m"].as<double>();
    }
    if (circuit["target_segment_length"]) {
      cfg.target_segment_length = circuit["target_segment_length"].as<double>();
    }
    if (circuit["samples_per_segment"]) {
      cfg.samples_per_segment = circuit["samples_per_segment"].as<int>();
    }
  }
  return cfg;
}

std::string MpccController::rekeyMpccToMpc(const std::string& source_path) {
  const YAML::Node root = YAML::LoadFile(source_path);
  if (!root.IsMap() || !root["mpcc"]) {
    return source_path;
  }
  // Build a copy with mpcc: renamed to mpc:.
  YAML::Node rekeyed;
  for (auto it = root.begin(); it != root.end(); ++it) {
    const std::string key = it->first.as<std::string>();
    rekeyed[(key == "mpcc") ? "mpc" : key] = it->second;
  }

  // Write the temp file under $TMPDIR (or /tmp). Use mkstemp for a
  // unique name; the upstream loader only reads it once at startup so
  // it is safe to leave for the OS to reap.
  std::filesystem::path tmpdir =
      std::filesystem::temp_directory_path();  // honours TMPDIR
  std::string templ =
      (tmpdir / "mpcc_rekey_XXXXXX.yaml").string();
  std::vector<char> buf(templ.begin(), templ.end());
  buf.push_back('\0');
  const int fd = mkstemps(buf.data(), 5);  // suffix length: ".yaml" == 5
  if (fd < 0) {
    throw std::runtime_error("MpccController::rekeyMpccToMpc: mkstemps failed.");
  }
  const std::string tmp_path(buf.data());
  ::close(fd);

  std::ofstream out(tmp_path);
  if (!out) {
    throw std::runtime_error("MpccController::rekeyMpccToMpc: cannot open " + tmp_path);
  }
  out << "# AUTO-GENERATED by MpccController::rekeyMpccToMpc — do not edit by hand.\n";
  out << rekeyed;
  out.close();
  return tmp_path;
}

void MpccController::initialize(const mav_model::State& initial_state,
                                const ExampleConfig& example_cfg) {
  if (example_cfg.mpc_dt <= 0.0) {
    throw std::invalid_argument("MpccController: example_cfg.mpc_dt must be > 0.");
  }

  // MPCC solver + YAML re-key shim. `configureMpcFromYaml` upstream
  // reads root['mpc'] hardcoded; our YAMLs use `mpcc:` so we rewrite
  // to a temp file before invoking the loader.
  mpc_ = std::make_unique<acados_mpc::MPC>();
  const std::string rekeyed_yaml = rekeyMpccToMpc(cfg_.mpc_yaml_path);
  acados_mpc::configureMpcFromYaml(*mpc_, rekeyed_yaml);
  // Note: ocp_json_file_path is currently informational — the MPC
  // constructor loads the codegen .so via RUNPATH baked at build time,
  // not from the JSON. We still validate the json exists for parity
  // with the Python adapter and to fail fast if the user mis-configured.
  (void)resolveOcpJson(cfg_.ocp_json_file_path, cfg_.mpc_yaml_path);

  control_period_ = mpc_->getPredictionTimeStep();
  if (example_cfg.mpc_dt > 0.0) {
    control_period_ = example_cfg.mpc_dt;
  }

  // Mission + spline. Path precedence: env var > config YAML > demo defaults.
  const std::string env_mission = envOrEmpty("CIRCUIT_MISSION_YAML");
  const std::string env_gates   = envOrEmpty("CIRCUIT_GATES_YAML");
  const auto [demo_mission, demo_gates] = ml::defaultPaths();
  const std::string mission_yaml =
      !env_mission.empty() ? env_mission
                           : (!cfg_.mission_yaml.empty() ? cfg_.mission_yaml : demo_mission);
  const std::string gates_yaml =
      !env_gates.empty() ? env_gates
                         : (!cfg_.gates_yaml.empty() ? cfg_.gates_yaml : demo_gates);

  Eigen::Vector3d origin_pose = initial_state.getPositionVector();
  if (example_cfg.takeoff_altitude_m > 0.0) {
    origin_pose.z() = example_cfg.takeoff_altitude_m;
  }
  const auto setpoints = CircuitGenerator::buildSetpoints(
      origin_pose, mission_yaml, gates_yaml, cfg_.origin_offset_m, cfg_.closing_exit_margin_m);
  traj_ = std::make_unique<spline::TrajectoryGenerator>(
      setpoints, kMpccSplineKnots, cfg_.target_segment_length, cfg_.samples_per_segment);

  s_eval_                = 0.0;
  last_solve_us_         = 0.0;
  last_desired_velocity_ = Eigen::Vector3d::Zero();
}

framework::ControlCommand MpccController::computeCommand(
    const mav_model::State& state,
    const std::vector<framework::ReferenceSample>& /*references*/) {
  if (!mpc_ || !traj_) {
    throw std::runtime_error("MpccController: initialize() must be called first.");
  }
  const Eigen::Vector3d pos       = state.getPositionVector();
  const Eigen::Quaterniond q      = state.getOrientationVector();
  const Eigen::Vector3d vel       = state.getLinearVelocityVector();

  acados_mpc::MPCData* mpc_data = mpc_->getData();
  mpc_data->state.setPosition({pos.x(), pos.y(), pos.z()});
  mpc_data->state.setOrientation({q.w(), q.x(), q.y(), q.z()});
  mpc_data->state.setLinearVelocity({vel.x(), vel.y(), vel.z()});
  mpc_data->state.setTheta(s_eval_);

  auto [win, s_eval] = traj_->getTrajectoryWindow(s_eval_);
  s_eval_            = s_eval;

  // Pack the sliding window into the solver's spline parameters.
  std::array<double, kMpccSplineKnots * 3> spline_pts{};
  std::array<double, kMpccSplineKnots * 3> spline_tan{};
  std::array<double, kMpccSplineKnots * 3> spline_face{};
  std::array<double, kMpccSplineKnots>     s_lengths{};
  std::array<double, kMpccSplineKnots>     softc{};
  const std::size_t n_pts = std::min(win.path.position.size() / 3,
                                      static_cast<std::size_t>(kMpccSplineKnots));
  for (std::size_t i = 0; i < n_pts; ++i) {
    spline_pts[3 * i + 0] = win.path.position[3 * i + 0];
    spline_pts[3 * i + 1] = win.path.position[3 * i + 1];
    spline_pts[3 * i + 2] = win.path.position[3 * i + 2];
    spline_tan[3 * i + 0] = win.path.tangent[3 * i + 0];
    spline_tan[3 * i + 1] = win.path.tangent[3 * i + 1];
    spline_tan[3 * i + 2] = win.path.tangent[3 * i + 2];
    spline_face[3 * i + 0] = win.facing_points[3 * i + 0];
    spline_face[3 * i + 1] = win.facing_points[3 * i + 1];
    spline_face[3 * i + 2] = win.facing_points[3 * i + 2];
  }
  const std::size_t n_s = std::min(win.s_values.size(),
                                    static_cast<std::size_t>(kMpccSplineKnots));
  for (std::size_t i = 0; i < n_s; ++i) {
    s_lengths[i] = win.s_values[i];
  }
  softc.fill(kSoftWeightDefault);
  for (const auto& og : win.original_points) {
    if (og.index >= 0 && og.index < kMpccSplineKnots) {
      softc[og.index] = kSoftWeightOg;
    }
  }
  acados_mpc::OnlineParameters* params = mpc_->getParameters();
  params->setSplinePoints(spline_pts);
  params->setSplineTangents(spline_tan);
  params->setSplineFacePoints(spline_face);
  params->setSplineSoftconstraints(softc);
  params->setSLengths(s_lengths);

  // Solve.
  const auto t0    = std::chrono::high_resolution_clock::now();
  const int status = mpc_->solve();
  const auto t1    = std::chrono::high_resolution_clock::now();
  last_solve_us_   = std::chrono::duration<double>(t1 - t0).count() * 1e6;
  if (status != 0) {
    throw std::runtime_error("MpccController: solver returned status " +
                             std::to_string(status));
  }

  // Project the drone onto the spline (monotonic) for the next iteration.
  if (!win.s_values.empty()) {
    const double s_top   = win.s_values.back();
    const double s_lower = std::max(0.0, s_eval_ - kSearchRadiusM);
    const double s_upper = std::min(s_top, s_eval_ + kSearchRadiusM);
    double best_s        = findClosestS(pos, win, s_lower, s_upper);
    best_s               = std::max(s_eval_, best_s);
    s_eval_              = std::min(std::max(best_s, 0.0), s_top);
  }

  // Expose stage-1 predicted velocity (best proxy for v_des).
  if (win.path.tangent.size() >= 3) {
    last_desired_velocity_ = cfg_.desired_speed * Eigen::Vector3d(win.path.tangent[0],
                                                                  win.path.tangent[1],
                                                                  win.path.tangent[2]);
  }

  framework::ControlCommand cmd;
  cmd.thrust_n      = mpc_data->actuation.getThrust();
  const auto rates  = mpc_data->actuation.getAngularVelocity();
  cmd.angular_rate  = {rates[0], rates[1], rates[2]};
  return cmd;
}

}  // namespace mpc_examples::adapters
